python_modules/dagster/dagster/core/definitions/repository.py

```diff
@@ -594,68 +621,129 @@ def from_list(cls, repository_definitions):
                     partition_sets[partition_set_def.name] = partition_set_def
             elif isinstance(definition, GraphDefinition):
                 coerced = definition.coerce_to_job()
-                pipelines[coerced.name] = coerced
+                if coerced.name in pipelines_and_jobs:
+                    raise DagsterInvalidDefinitionError(
+                        "Duplicate {target_type} definition found for {target}".format(
+                            target_type=definition.target_type, target=definition.describe_target()
+                        )
+                    )
+                pipelines_and_jobs[coerced.name] = coerced
             else:
                 check.failed(f"Unexpected repository entry {definition}")
 
         return CachingRepositoryData(
-            pipelines=pipelines,
+            pipelines_and_jobs=pipelines_and_jobs,
             partition_sets=partition_sets,
             schedules=schedules,
             sensors=sensors,
         )
 
     def get_pipeline_names(self):
-        """Get the names of all pipelines in the repository.
+        """Get the names of all pipelines/jobs in the repository.
 
         Returns:
             List[str]
         """
-        return self._pipelines.get_definition_names()
+        return self._pipelines_and_jobs.get_definition_names()
+
+    def get_job_names(self):
+        """Get the names of all jobs in the repository.
+
+        Returns:
+            List[str]
+        """
+        return [job.name for job in self.get_all_jobs()]
 
     def has_pipeline(self, pipeline_name):
-        """Check if a pipeline with a given name is present in the repository.
+        """Check if a pipeline/job with a given name is present in the repository.
 
         Args:
-            pipeline_name (str): The name of the pipeline.
+            pipeline_name (str): The name of the pipeline/job.
 
         Returns:
             bool
         """
         check.str_param(pipeline_name, "pipeline_name")
-        return self._pipelines.has_definition(pipeline_name)
+
+        return self._pipelines_and_jobs.has_definition(pipeline_name)
+
+    def has_job(self, job_name):
+        """Check if a job with a given name is present in the repository.
+
+        Args:
+            job_name (str): The name of the job.
+
+        Returns:
+            bool
+        """
+        check.str_param(job_name, "job_name")
+
+        if self._pipelines_and_jobs.has_definition(job_name):
+            if isinstance(self.get_pipeline(job_name), JobDefinition):
```

Comment:
I think this is the main issue w/ keeping pipelines_and_jobs together under a single key. The old implementation did not have to instantiate the pipeline in order to tell if the pipeline existed or not.

Reverting back to having them in separate caches should fix this issue.